js/plugins/vertexai/src/vector-search/indexers.ts

```diff
@@ -0,0 +1,116 @@
+/**
+ * Copyright 2024 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { embedMany } from '@genkit-ai/ai/embedder';
+
+import {
+  defineIndexer,
+  IndexerAction,
+  indexerRef,
+} from '@genkit-ai/ai/retriever';
+
+import { logger } from '@genkit-ai/core/logging';
+import z from 'zod';
+import {
+  Datapoint,
+  vertexVectorSearchOptions,
+  VVSIndexerOptionsSchema,
+} from './types';
+import { upsertDatapoints } from './upsert_datapoints';
+
+export const vertexAiIndexerRef = (params: {
+  indexId: string;
+  displayName?: string;
+}) => {
+  return indexerRef({
+    name: `vertexai/${params.indexId}`,
+    info: {
+      label: params.displayName ?? `vertexAi - ${params.indexId}`,
+    },
+    configSchema: z.any().optional(),
+  });
+};
+
+export function vertexIndexers<EmbedderCustomOptions extends z.ZodTypeAny>(
+  params: vertexVectorSearchOptions<EmbedderCustomOptions>
+) {
+  const vectorSearchOptions = params.pluginOptions.vectorSearchIndexOptions!;
```

Comment:
Is there anyway to avoid the `tailbang` operator here?

Reply:
ah this is why that MakeRequired type was hanging around haha, i'll have a think, we can def avoid this